doc/source/serve/production-guide/fault-tolerance.md

```diff
@@ -0,0 +1,273 @@
+(serve-e2e-ft)=
+# Adding End-to-End Fault Tolerance
+
+This section helps you:
+
+* provide additional fault tolerance for your Serve application
+* understand Serve's recovery procedures
+* simulate system errors in your Serve application
+
+:::{admonition} Relevant Guides
+:class: seealso
+This section discusses concepts from:
+* Serve's [architecture guide](serve-architecture)
+* Serve's [Kubernetes production guide](serve-in-production-kubernetes)
+:::
+
+(serve-e2e-ft-guide)=
+## Guide: end-to-end fault tolerance for your Serve app
+
+Serve provides some [fault tolerance](serve-ft-detail) features out of the box. You can provide end-to-end fault tolerance by tuning these features and running Serve on top of [KubeRay].
+
+### Replica health-checking
+
+By default, the Serve controller periodically health-checks each Serve deployment replica and restarts it on failure.
+
+You can define custom application-level health-checks and adjust their frequency and timeout.
+To define a custom health-check, add a `check_health` method to your deployment class.
+This method should take no arguments and return no result, and it should raise an exception if the replica should be considered unhealthy.
+The Serve controller logs the raised exception if the health-check fails.
+You can also use the deployment options to customize how frequently the health-check is run and the timeout after which a replica is marked unhealthy.
+
+```{literalinclude} ../doc_code/fault_tolerance/replica_health_check.py
+:start-after: __health_check_start__
+:end-before: __health_check_end__
+:language: python
+```
+
+### Worker node recovery
+
+:::{admonition} KubeRay Required
+:class: caution, dropdown
+You **must** deploy your Serve application with [KubeRay] to use this feature.
+
+See Serve's [Kubernetes production guide](serve-in-production-kubernetes) to learn how you can deploy your app with KubeRay.
+:::
+
+By default, Serve can recover from certain failures, such as unhealthy actors. When [Serve runs on Kubernetes](serve-in-production-kubernetes) with [KubeRay], it can also recover from some cluster-level failures, such as dead worker or head nodes.
+
+When a worker node fails, the actors running on it also fail. Serve detects that the actors have failed, and it attempts to respawn the actors on the remaining, healthy nodes. Meanwhile, KubeRay detects that the node itself has failed, so it brings up a new healthy node to replace it. Serve can then respawn any pending actors on that node as well. The deployment replicas running on healthy nodes can continue serving traffic throughout the recovery period.
+
+(serve-e2e-ft-guide-gcs)=
+### Head node recovery: Ray GCS fault tolerance
```

Comment:
Reading the doc, quick question from my mind, is the redis information always synced? Will there be gap of cluster between failure states and recover states?

Reply:
Good question, @DmitriGekhtman @brucez-anyscale do you know how this case is handled?